### ISSUE TITLE
[Proposal] infer Query Builder join columns

### ISSUE THREAD

**Issue by brynj-digital** [laravel/internals#931]
This would be a nice inclusion if the query builder inferred the joining columns when doing joins, to increase the brevity. This would also mirror Eloquent behaviour.

For example, using the example in the docs ([https://laravel.com/docs/5.5/queries#joins](https://laravel.com/docs/5.5/queries#joins)) this:

    $users = DB::table('users')
        ->join('contacts', 'users.id', '=', 'contacts.user_id')
        ->join('orders', 'users.id', '=', 'orders.user_id')
        ->select('users.*', 'contacts.phone', 'orders.price')
        ->get();

....would become this:

    $users = DB::table('users')
        ->join('contacts')
        ->join('orders')
        ->select('users.*', 'contacts.phone', 'orders.price')
        ->get();

Overrides would still be possible of course (and necessary in many instances), but it would be a non-breaking change.

**Comment by mfn** [comment]
I like that.

I write a lot of `Builder` joins and in reality the names of tables are usually longer and for 99% of the time it's this case.

Thinking about an implementation:
- how would one get from `users` to `user`? Yeah, sure we strip the `s` but, will this always work? Is it enough to use `\Illuminate\Support\Pluralizer::singular`? 🤷‍♀️
- I'm also interested if a solution would only play around with the strings _without_ having to know meta-data about the actual tables to infer this?

**Comment by 36864** [comment]
I'd expect the implementation to match the logic of inferring relationship columns.
It won't _always_ work, but it will work in the trivial case, which is all that we can realistically hope for, just like with relationships.

**Comment by danjdewhurst** [comment]
Exactly what @36864 said. I don't see a reason to defer away from the current standards with Eloquent.

**Comment by mfn** [comment]
> from the way Eloquent currently works.

But do we have this information?

Eloquent has more information available due to the underlying models. It has access to facts like `getForeignKey()` or `getKeyName()` which the `Builder` hasn't AFAIK.

<details>
<summary>I did a small PoC, at least no tests break</summary>

```diff
diff --git src/Illuminate/Database/Query/Builder.php src/Illuminate/Database/Query/Builder.php
index 29c0e1f27..f2056a99a 100755
--- src/Illuminate/Database/Query/Builder.php
+++ src/Illuminate/Database/Query/Builder.php
@@ -333,14 +333,14 @@ class Builder
      * Add a join clause to the query.
      *
      * @param  string  $table
-     * @param  string  $first
+     * @param  string|null  $first
      * @param  string|null  $operator
      * @param  string|null  $second
      * @param  string  $type
      * @param  bool    $where
      * @return $this
      */
-    public function join($table, $first, $operator = null, $second = null, $type = 'inner', $where = false)
+    public function join($table, $first = null, $operator = null, $second = null, $type = 'inner', $where = false)
     {
         $join = new JoinClause($this, $type, $table);

diff --git src/Illuminate/Database/Query/JoinClause.php src/Illuminate/Database/Query/JoinClause.php
index 4b32df29e..23f189340 100755
--- src/Illuminate/Database/Query/JoinClause.php
+++ src/Illuminate/Database/Query/JoinClause.php
@@ -3,6 +3,7 @@
 namespace Illuminate\Database\Query;

 use Closure;
+use Illuminate\Support\Pluralizer;

 class JoinClause extends Builder
 {
@@ -58,7 +59,7 @@ class JoinClause extends Builder
      *
      * on `contacts`.`user_id` = `users`.`id`  and `contacts`.`info_id` = `info`.`id`
      *
-     * @param  \Closure|string  $first
+     * @param  \Closure|string|null $first
      * @param  string|null  $operator
      * @param  string|null  $second
      * @param  string  $boolean
@@ -66,12 +67,21 @@ class JoinClause extends Builder
      *
      * @throws \InvalidArgumentException
      */
-    public function on($first, $operator = null, $second = null, $boolean = 'and')
+    public function on($first = null, $operator = null, $second = null, $boolean = 'and')
     {
         if ($first instanceof Closure) {
             return $this->whereNested($first, $boolean);
         }

+        // If we're not provided any columns we assume a "classic" join
+        if (!$first && !$second) {
+            $first = $this->table.'.id';
+
+            $operator = '=';
+
+            $second = $this->parentQuery->from.'.'.Pluralizer::singular($this->table).'_id';
+        }
+
         return $this->whereColumn($first, $operator, $second, $boolean);
     }
```
</details>

This gets us so far but as others said, not the wanted way to approach this:
```sh
$ ./artisan tinker
Psy Shell v0.8.16 (PHP 7.1.12 — cli) by Justin Hileman
>>> DB::table('posts')->join('users')->toSql();
=> "select * from "posts" inner join "users" on "users"."id" = "posts"."user_id""
```

**Comment by 36864** [comment]
>Eloquent has more information available due to the underlying models. It has access to facts like getForeignKey() or getKeyName() which the Builder hasn't AFAIK.

You're right, I didn't consider that. What I meant was the way Eloquent works if everything is left as default (which I now realize is probably just a pluralizer call but I haven't checked).

**Comment by sisve** [comment]
We cannot use `Pluralizer::singular` because we accept PRs all the time that changes which words are uncountable. I once built code that attempted this; and it failed when a change to the uncountable list broke my database calls. https://github.com/laravel/framework/issues/6675

**Comment by 36864** [comment]
Had some time to look at how Eloquent does this and it uses class names to get default foreign key names, which we wouldn't have access to.
I don't see a way to do this without access to model instances or pluralizer.
Unless changes to $uncountable become regarded as BC breaks (which they possibly should be), I don't think this will be possible except maybe for just the trivial case of 's' pluralization being hard-coded into the method. I don't know if that would be worth the trouble.

**Comment by brynj-digital** [comment]
> We cannot use Pluralizer::singular because we accept PRs all the time that changes which words are uncountable. 

> Had some time to look at how Eloquent does this and it uses class names to get default foreign key names, which we wouldn't have access to.
I don't see a way to do this without access to model instances or pluralizer.

That is a real shame, it would have been a nice feature to include.

**Comment by mfn** [comment]
Then my suggesting would be: do the most common use-case thing and lets do a `rtim('s')`.

My arguments would be:
- still works in 99% cases
- will never "break" (see the Pluralizer issue, which I totally get)
- your referenced table doesn't end in an `s`? Too bad, then the short-cut improvement isn't for you and you have to use the full syntax like you always did. Nothing won, but also nothing lost.

It doesn't have to be perfect; it has to be usable. How does that sound?

**Comment by brynj-digital** [comment]
@mfn I would be onboard with this - zero intelligence, zero risk.

**Comment by mfn** [comment]
PR opened at https://github.com/laravel/framework/pull/22480

**Comment by brynj-digital** [comment]
thanks @mfn, hope the PR is accepted :thumbsup:

**Comment by fico7489** [comment]
see https://github.com/fico7489/laravel-eloquent-join
